src/Serilog.Sinks.Datadog.Logs/Sinks/Datadog/DatadogHttpClient.cs

```diff
@@ -36,123 +42,141 @@ public class DatadogHttpClient : IDatadogClient
             _client = new HttpClient();
+            _client.BaseAddress = new Uri(config.Url);
             _client.DefaultRequestHeaders.Add("DD-API-KEY", apiKey);
             _client.DefaultRequestHeaders.Add("DD-EVP-ORIGIN", "Serilog.Sinks.Datadog.Logs");
             _client.DefaultRequestHeaders.Add("DD-EVP-ORIGIN-VERSION", _version);
-            _url = $"{config.Url}/api/v2/logs";
             _formatter = formatter;
+            _recycleResources = recycleResources;
         }
 
-        public Task WriteAsync(IEnumerable<LogEvent> events)
+        /// <summary>
+        /// Writes the specified <paramref name="events"/> to Datadog logs route. 
+        /// </summary>
+        /// <param name="events">The events that will be sent.</param>
+        /// <param name="onException">The callback that is invoked when an exception occurs in this event but is handled gracefully.</param>
+        /// <remarks>
+        /// <para>This method uses a semaphore so that only one batch of events is ever on the wire at any given time.</para>
+        /// <para>This allows for a single allocated contiguous memory to be written to for serialization of events.</para>
+        /// </remarks>
+        public async Task WriteAsync(LogEvent[] events, Action<Exception> onException)
         {
-            var serializedEvents = SerializeEvents(events);
-            var tasks = serializedEvents.LogEventChunks.Select(Post);
-
-            var tooBigTask = Task.Run(() =>
+            var builder = _recycleResources ? _chunkBuilder : new StringBuilder();
+            try
             {
-                if (serializedEvents.TooBigLogEvents.Count > 0)
-                {
-                    throw new TooBigLogEventException(serializedEvents.TooBigLogEvents);
-                }
-            });
+                var chunkCount = 0;
+                var chunkStart = 0;
 
-            tasks = tasks.Concat(new[] { tooBigTask });
-
-            return Task.WhenAll(tasks);
-        }
-
-        private SerializedEvents SerializeEvents(IEnumerable<LogEvent> events)
-        {
-            var serializedEvents = new SerializedEvents();
-            int currentSize = 0;
+                for (var i = 0; i < events.Length; i++)
+                {
 
-            var chunkBuffer = new List<string>(events.Count());
-            var logEvents = new List<LogEvent>(events.Count());
-            foreach (var logEvent in events)
-            {
-                var formattedLog = _formatter.FormatMessage(logEvent);
-                var logSize = Encoding.UTF8.GetByteCount(formattedLog);
-                if (logSize > _maxMessageSize)
+                    if (_cancellationToken.IsCancellationRequested)
+                    {
+                        break;
+                    }
+                    var formattedLog = _formatter.FormatMessage(events[i]);
+                    var logSize = Encoding.UTF8.GetMaxByteCount(formattedLog.Length);
```

Comment:
I'll switch this back. Because we work primarily with ASCII data `GetMaxByteCount` is usually safe, but I realize that isn't going to be the case for everyone.